packages/headless/src/controllers/triggers/headless-query-trigger.ts

```diff
@@ -0,0 +1,82 @@
+import {Engine} from '../../app/headless-engine';
+import {ConfigurationSection, TriggerSection} from '../../state/state-sections';
+import {configuration, triggers} from '../../app/reducers';
+import {buildController, Controller} from '../controller/headless-controller';
+import {loadReducerError} from '../../utils/errors';
+import {executeSearch} from '../../features/search/search-actions';
+import {
+  updateQuery,
+  UpdateQueryActionCreatorPayload,
+} from '../../features/query/query-actions';
+import {logTriggerQuery} from '../../features/triggers/trigger-analytics-actions';
+
+/**
+ * The `QueryTrigger` controller handles query triggers.
+ */
+export interface QueryTrigger extends Controller {
+  /**
+   * the state of the `QueryTrigger` controller.
+   */
+  state: QueryTriggerState;
+}
+
+/**
+ * A scoped and simplified part of the headless state that is relevant to the `QueryTrigger` controller.
+ */
+export interface QueryTriggerState {
+  /**
+   * The new query to perform a search with after receiving a query trigger.
+   */
+  query: string | null;
+}
+
+/**
+ * Creates a `QueryTrigger` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @returns A `QueryTrigger` controller instance.
+ * */
+export function buildQueryTrigger(engine: Engine<object>): QueryTrigger {
+  if (!loadQueryTriggerReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+
+  const getState = () => engine.state;
+
+  return {
+    ...controller,
+
+    subscribe(listener: () => void) {
+      const strictListener = () => {
+        if (this.state.query) {
+          listener();
+          const updateQueryPayload: UpdateQueryActionCreatorPayload = {
+            q: this.state.query,
+          };
+          dispatch(updateQuery(updateQueryPayload));
+          //maybe logSearchBoxSubmit instead
+          //and dispatch logTriggerQuery  separately?
+          dispatch(executeSearch(logTriggerQuery()));
```

Comment:
Not 100% sure this was the right way to update the query and perform a new search. When testing manually with the react samples, the query successfully updates because I can see the value of `q` change on the url. However, the page does not rerender with the results from the new query since the new executeSearch action with the new query is never fulfilled due to `Action dispatch error query/UpdateQuery`

Reply:
Is there a message in the network tab of the chrome dev tools perhaps?

Another thought, when the executeSearch is dispatched, are we clearing the query trigger from state to avoid an infinite loop?

Reply:
That is what I was thinking as well but I thought the new executeSearch action would fulfill and if the response does not contain any query triggers, then the controller's `state.query` would be set to an empty string and not result in an infinite loop. But I think your suggestion might work.